lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp

```diff
@@ -306,13 +352,51 @@ SILValue PullbackEmitter::getAdjointProjection(SILBasicBlock *origBB,
     return builder.createTupleElementAddr(teai->getLoc(), adjSource, adjIndex);
   }
   // Handle `begin_access`.
+  // Adjoint projection: the base adjoint buffer itself.
   if (auto *bai = dyn_cast<BeginAccessInst>(originalProjection)) {
     auto adjBase = getAdjointBuffer(origBB, bai->getOperand());
     if (errorOccurred)
       return (bufferMap[{origBB, originalProjection}] = SILValue());
     // Return the base buffer's adjoint buffer.
     return adjBase;
   }
+  // Handle `array.uninitialized_intrinsic` application element addresses.
```

Comment:
This PR introduces two main changes to pullback generation:
1. Adds a case to `PullbackEmitter::getAdjointProjection` for array literal element addresses.
2. Adds special logic to `PullbackEmitter::addAdjointValue`: when adding adjoint value for array literal, accumulate also for its element addresses.

With (1) and (2), TF-975 and TF-978 tests are fixed. I wondered if (1) can be dropped: perhaps array literal element addresses don't need special adjoint buffer initialization, and special logic for accumulating element address adjoint buffers in `PullbackEmitter::addAdjointValue` or `PullbackEmitter::setAdjointValue` is sufficient.

---

Here are the results:

Dropping (1) and keeping (2) for `PullbackEmitter::addAdjointValue` leads to correctness issues:
```
[ RUN      ] ArrayAutoDiff.ArrayIdentity
[       OK ] ArrayAutoDiff.ArrayIdentity
[ RUN      ] ArrayAutoDiff.ArraySubscript
[       OK ] ArrayAutoDiff.ArraySubscript
[ RUN      ] ArrayAutoDiff.ArrayLiteral
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 38
stdout>>> expected: Tracked(1.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 38
stdout>>> expected: Tracked(2.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 49
stdout>>> expected: Tracked(8.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 49
stdout>>> expected: Tracked(6.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 58
stdout>>> expected: Tracked(8.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 58
stdout>>> expected: Tracked(6.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 68
stdout>>> expected: Tracked(8.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> check failed at /Users/danielzheng/swift-bart/swift/test/AutoDiff/array.swift, line 68
stdout>>> expected: Tracked(6.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
stdout>>> actual: Tracked(0.0) (of type DifferentiationUnittest.Tracked<Swift.Float>)
[     FAIL ] ArrayAutoDiff.ArrayLiteral
...
```
This may be because `PullbackEmitter::addAdjointValue` isn't always called for array literal values. `PullbackEmitter::setAdjointValue` should always be called, though.

---

Dropping (1) and keeping (2) for `PullbackEmitter::setAdjointValue` leads to `SIL verification failed: Basic blocks must end with a terminator instruction: isa<TermInst>(BB.back())` issues: https://gist.github.com/dan-zheng/504a2ab65f8b999b8ce9478854d5f46b

The insertion point of `builder` in `PullbackEmitter::accumulateArrayLiteralElementAddressAdjoints` seems sensitive. Perhaps it can be massaged in a way that fixes the SIL verification errors while enabling (1) to be dropped.